apollo-federation/src/operation/optimize.rs

```diff
@@ -1403,19 +1406,12 @@ impl SelectionSet {
         // the named fragment, and in that case we return a singleton selection with just that.
         // Otherwise, it's our wrapping inline fragment with the sub-selections optimized, and we
         // just return that subselection.
-        match optimized {
-            InlineOrFragmentSelection::FragmentSpread(_) => {
-                let self_selections = Arc::make_mut(&mut self.selections);
-                self_selections.clear();
-                self_selections.insert(optimized.into());
-            }
-
-            InlineOrFragmentSelection::InlineFragment(inline_fragment) => {
-                // Note: `inline_fragment.selection_set` can't be moved (since it's inside Arc).
-                // So, it's cloned.
-                *self = inline_fragment.selection_set.clone();
+        *self = match optimized {
```

Comment:
Replacing the whole selection set avoids potentially cloning a big `self.selections` map only to immediately clear it.